### DISCUSSION TITLE
Use returned object to generate cache id

### DISCUSSION THREAD

**Discussion by dunklesToast** [mikro-orm/mikro-orm#3256] [Ideas]
**Is your feature request related to a problem? Please describe.**
I sometimes have a query where it's possible to find one specific result in the database with multiple queries.
Let's say this is our element:
``` json
{
  "id": "ZDLLqbmN",
  "name": "dunklesToast",
  "some": "value"
}
```

We could now fetch this user like this
```ts
return this.entityManager
  .createQueryBuilder(UserEntity, 'u')
  .where({ 'u.id': identifier })
  .orWhere({ 'u.name': identifier })
  .getSingleResult();
```
This query works awesome unless you add a cache:
```ts
return this.entityManager
  .createQueryBuilder(UserEntity, 'u')
  .where({ 'u.id': identifier })
  .orWhere({ 'u.name': identifier })
  .cache(['user-' + identifier, 180])
  .getSingleResult();
```
Since the `identifier` could be either `dunklesToast` or `ZDLLqbmN` I'll end up with multiple cache elements which introduces some issues such as making purging the cache more complicated.

**Describe the solution you'd like**
Having the possibility to create the cache key based on a value of the returned object would be nice. I am not entirely, how to do a great markup for this, but we might come up with something great. One way would be to use Handlebars to support a syntax like shown in [this demo](https://handlebarsjs.com/playground.html#format=1&currentExample=%7B%22template%22%3A%22user-%7B%7Bid%7D%7D%5Cn%22%2C%22partials%22%3A%5B%5D%2C%22input%22%3A%22%7B%5Cn%20%20%5C%22id%5C%22%3A%20%5C%22ZDLLqbmN%5C%22%2C%5Cn%20%20%5C%22name%5C%22%3A%20%5C%22dunklesToast%5C%22%20%2C%5Cn%20%20%5C%22some%5C%22%3A%20%5C%22value%5C%22%5Cn%7D%22%2C%22output%22%3A%22user-ZDLLqbmN%5Cn%22%2C%22preparationScript%22%3A%22%5Cn%22%2C%22handlebarsVersion%22%3A%224.7.7%22%7D), where you can use `user-{{id}}` to access the ID of the user. This would always result in the same cache key no matter which identifier is used.

As far as I know this behavior just needs to be implemented directly to the cache adapters so it should be a pretty straight forward thing to do. If you also think that this is a nice addition to MikroORM I could open a PR for this.

**Describe alternatives you've considered**
I could just create my own CacheAdapter which would support this but I thought it would be great to put this out for everyone.

**Comment by B4nan** [comment]
How could you have a cache key that depend on the value? How can you know there is a cache hit without actually performing the query? How can we even know what should be the cache key based on? You dont provide any `UserEntity.id` anywhere in your code, that QB instance is not aware of such value.

If you know how to do this, feel free to propose via PR, as here I have no idea what is the end goal (although can't really promise this would be merged). You dont like having multiple cache elements, but you need them, as you need to cache different queries with different responses, or not? Your QB is not aware of any identifier, and I parsing the query sounds like a very bad idea to me (not really interested in that, anywhere, not just cache related, parsing queries is for database engine, not for ORMs).

**Reply by dunklesToast** [reply]
Oh yea. I had a brain lag there. I honestly just thought about writing to the cache (which would be easily be possible since you've already got the data) but yea, reading from the cache is a different story and would probably not work in an easy way. Sorry for bothering!
